---
title: "Class 10 Lab"
author: "Cindy Tran"
date: "2/8/2022"
output: pdf_document
---

# Section 4: Population Scale Analysis

One sample is obviously not enough to know what is happening in a population. You are interested in assessing genetic differences on a population scale. 

So, you processed about ~230 samples and did the normalization on a genome level. Now, you want to find whether there is any association of the 4 asthma-associated SNPs (rs8067378…) on ORMDL3 expression.

How many samples do we have?

```{r}
expr <- read.table("rs8067378_ENSG00000172057.6.txt")
head(expr)
```

```{r}
nrow(expr)
```

> Q13: Read this file into R and determine the sample size for each genotype and their corresponding median expression levels for each of these genotypes. 

Hint: The read.table(), summary() and boxplot() functions will likely be useful here. There is an example R script online to be used ONLY if you are struggling in vein. Note that you can find the medium value from saving the output of the boxplot() function to an R object and examining this object. There is also the medium() and summary() function that you can use to check your 
understanding.

```{r}
# Sample Sizes
table(expr$geno)
```

```{r}
# Median Expression Levels
medians <- aggregate(exp ~ geno, expr, median)
medians
```

> Q14: Generate a boxplot with a box per genotype, what could you infer from the relative expression value between A/A and G/G displayed in this plot? Does the SNP effect the 
expression of ORMDL3? 

Hint: An example boxplot is provided overleaf – yours does not need to be as polished as this one

```{r}
library(ggplot2)
```

Let's make a boxplot.

```{r}
boxplt <- ggplot(expr) +
  aes(x = geno, y = exp, fill = geno) +
  geom_boxplot(notch = TRUE)
boxplt
```

The expression value for A/A is significantly higher than the expression value for G/G. I can infer that having a G/G genotype results in a reduced expression of ORMDL3. 














