---
title: "Class 11 Structural Bioinformatics (Pt. 1)"
author: "Cindy Tran"
date: "1/22/2022"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1. Introduction to the RCSB Protein Data Bank

## PDB Statistics

Download a CSV file from the PDB site (accessible from “Analyze” > “PDB Statistics” > “by Experimental Method and Molecular Type”. Move this CSV file into your RStudio project and use it to answer the following questions:

```{r}
read.csv("Data Export Summary.csv")
```

> Q1. What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
(143950 + 8514 + 7610 + 2393 + 150 + 11) / (162665 + 9636 + 10014 + 3861 + 184 + 22) * 100

(6571 + 1086 + 2127 + 61 + 3) / (162665 + 9636 + 10014 + 3861 + 184 + 22) * 100
```
About 87% by X-ray and 5% by electron microscopy (about 92% by either X-ray or electron microscopy)

> Q2. What proportion of structures in the PDB are protein?

```{r}
162665 / (162665 + 9636 + 10014 + 3861 + 184 + 22) * 100
```
About 87% are protein only.

> Q3. Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

1856

# 2. Visualizing the HIV-1 Protease Structure

## Using Atom Selections

> Q4. Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

It is because we chose to represent the entire water molecule as one sphere. 

> Q5. There is a conserved water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have (see note below)?

127

## Sequence Viewer Extension [OPTIONAL]

> Q6. As you have hopefully observed HIV protease is a homodimer (i.e. it is composed of two identical chains). With the aid of the graphic display and the sequence viewer extension can you identify secondary structure elements that are likely to only form in the dimer rather than the monomer?

Coil

# 3. Introduction to Bio3D in R

```{r}
library(bio3d)
```

## Reading PDB File Data into R

```{r}
pdb <- read.pdb("1hsg")
pdb
```
> Q7. How many amino acid residues are there in this pdb object?

198

> Q8. Name one of the two non-protein residues? 

HOH

> Q9. How many protein chains are in this structure?

2

```{r}
attributes(pdb)
head(pdb$atom)
```

# 4. Comparative Structure Analysis of Adenylate Kinase

## Setup

```{r}
# Install packages in the R console not your Rmd

#install.packages("bio3d")
#install.packages("ggplot2")
#install.packages("ggrepel")
#install.packages("devtools")
#install.packages("BiocManager")

#BiocManager::install("msa")
#devtools::install_bitbucket("Grantlab/bio3d-view")
```

> Q10. Which of the packages above is found only on BioConductor and not CRAN?

msa

> Q11. Which of the above packages is not found on BioConductor or CRAN?

bio3d-view

> Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket?

True

## Search and Retrieve ADK Structures

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
aa
```

> Q13. How many amino acids are in this sequence, i.e. how long is this sequence?

214

```{r}
# Blast or hmmer search 
b <- blast.pdb(aa)
```

```{r}
# Plot a summary of search results
hits <- plot(b)
```

```{r}
# List out some 'top hits'
head(hits$pdb.id)
```

```{r}
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

## Align and Superpose Structures

```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```

```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
plot(pdbs, labels = ids)
```

## Optional: Viewing our superposed structures

```{r}
library(bio3d.view)
#install.packages("rgl")
library(rgl)

view.pdbs(pdbs)
```

## Principal Component Analysis

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

# 5. Optional Further Visualization

```{r}
# Visualize first principal component
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

```{r}
view.xyz(pc1)
```

```{r}
view.xyz(pc1, col=vec2color( rmsf(pc1) ))
```

```{r}
#Plotting results with ggplot2
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```

# 6. Normal Mode Analysis

```{r}
# NMA of all structures
modes <- nma(pdbs)
```

```{r}
plot(modes, pdbs, col=grps.rd)
```

> Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

The black and colored lines are similar in terms of overall shape, but they differ in that the black lines have less fluctuations overall than the colored lines. I think they differ most along residues 40-60 and residues 125-155. I think this is because these regions of residues are where the protein is most flexible and can change conformations. 























